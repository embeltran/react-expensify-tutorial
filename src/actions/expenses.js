import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});
// REMOVE_EXPENSE
export const startRemoveExpenses = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    }).catch((e) => {
      console.log('This failed:' , e );
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
// UPDATE EXPENSE
export const startEditExpenses = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense( id, updates ));
    }).catch((e) => {
      console.log('This failed:' , e );
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// fetch data
export const startSetExpenses = () => {
  return (dispatch) => {
    // connect to DB . fetch data 
    return database.ref('expenses').once('value').then((snapshot) =>{
      const expenses = [];
      //expenseSnapshot can also be childSnapshot === parse the data
      snapshot.forEach((expenseSnapshot) =>{
        expenses.push({
          id: expenseSnapshot.key, 
          ...expenseSnapshot.val() 
        });
      });
      //dispatch (show) setExpanses
      dispatch(setExpenses(expenses));

    });
  };
};
