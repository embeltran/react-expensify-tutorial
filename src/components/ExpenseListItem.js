import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <p>
      <strong><Link to={`/edit/${id}`}>{description}</Link></strong>
      {numeral(amount / 100).format('$0,0.00')}
      &nbsp;
      {moment(createdAt).format('YYYY-MM-DD')}
    </p>
);

export default ExpenseListItem;
