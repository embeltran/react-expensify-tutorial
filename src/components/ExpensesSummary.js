import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses' ;
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  
  return (
    <div className="expense-summary">
      <div>
        <h2>Viewing <strong>{expenseCount}</strong> <strong>{expenseWord}</strong> totalling <strong>{formattedExpensesTotal}</strong></h2>
        <p><Link to="/create" className="btn-expense">Add Expense</Link></p>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
