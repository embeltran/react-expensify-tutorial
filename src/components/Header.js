import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <p>
      <span>
        <NavLink to="/dashboard" activeClassName="is-active" exact={true}>Dashboard</NavLink>
      </span>
      &nbsp;
      <span>
        <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
      </span>
      &nbsp;
      <span>
        <NavLink to="/help" activeClassName="is-active">Help </NavLink>
      </span>
    </p>
  </header>
);

export default Header;
