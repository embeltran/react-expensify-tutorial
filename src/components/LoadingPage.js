import React from 'react';

const LoadingPage = () => (
    <div className="loader">
        <img src="/images/loader.gif" alt="" />
    </div>
);

export default LoadingPage;