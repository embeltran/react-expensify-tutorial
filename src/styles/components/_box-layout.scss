.box-layout {
    align-items: center;
    background:url(/images/bg.jpg) no-repeat;
    background-size: cover;
    display: flex;
    height: 100vh;
    justify-content: center;
    width: 100vw;
}
.box-layout > div {
    background: fade-out($color: white, $amount: .25);
    border-radius: 15px;
    text-align: center;
    padding: $l-size $m-size;
    width: 25rem;

}
.box-layout > div > h2 {
    text-transform: uppercase;
}
.box-layout > div > p {
    text-align: left;
}
.box-layout div > button {
    border-radius: 20px;
    padding: 4px 15px;
    font-weight: bold;
    line-height: 1;
    
}
.box-layout div > button:hover{
    color: #fff;
}
