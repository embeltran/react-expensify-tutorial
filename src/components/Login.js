import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const Login = ({ startLogin }) => (
  <div className="box-layout"> 
    <div>
      <h2>expensify app</h2>
      <p>tag line</p>      
      <button onClick={startLogin}>Login with Google</button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(Login);