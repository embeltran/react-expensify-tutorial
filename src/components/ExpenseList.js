import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div className="expense-item">
    <div className="expense-list-header">
      <div className="show-mobile">expenses</div>
      <div className="show-desktop">expense</div>
      <div className="show-desktop">amount</div>
      <div className="show-desktop">date</div>
    </div>
    {
      props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
          props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />;
          })
        )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
