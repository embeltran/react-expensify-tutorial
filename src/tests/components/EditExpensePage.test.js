import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { EditExpensePage } from '../../components/EditExpensePage';

let startEditExpenses, startRemoveExpenses, history, wrapper;

beforeEach(() => {
  startEditExpenses = jest.fn();
  startRemoveExpenses = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      startEditExpenses={startEditExpenses}
      startRemoveExpenses={startRemoveExpenses}
      history={history}
      expense={expenses[2]}
    />
  );
});

test('should render startEditExpense', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle startEditExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startEditExpenses).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
});

test('should handle startRemoveExpenses', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startRemoveExpenses).toHaveBeenLastCalledWith({
    id: expenses[2].id
  });
});
