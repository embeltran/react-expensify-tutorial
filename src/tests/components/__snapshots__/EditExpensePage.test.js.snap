// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render startEditExpense 1`] = `
<div
  className="expense-edit"
>
  <div>
    <div
      className="content-container"
    >
      <h2>
        Edit Expense
      </h2>
    </div>
  </div>
  <div
    className="editExpenseForm"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 4500,
          "createdAt": 345600000,
          "description": "Credit Card",
          "id": "3",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
